mypy/types.py

```diff
@@ -1741,7 +1747,9 @@ def set_typ_args(tp: Type, new_args: List[Type], line: int = -1, column: int = -
     if isinstance(tp, Instance):
         return Instance(tp.type, new_args, line, column)
     if isinstance(tp, TupleType):
-        return tp.copy_modified(items=new_args)
+        fallback_args = [UnionType.make_simplified_union(new_args)]
```

Comment:
There is an item in the roadmap about documenting the details of different passes.
(Also there is an item about re-working semantic analysis passes.)